### ISSUE TITLE
How to optimize iOS and Android (Net 6) for package size and startup speed (Trimmer and Profile Guided AOT)

### ISSUE THREAD

**Issue by jeanplevesque** [unoplatform/uno#11248]
## What would you like clarification on:

Now that Xamarin is phasing out, there should be a guide on how to optimize Uno applications with the unified .Net options, especially on mobile platforms (iOS & Android).

There are at least 2 areas of interest.

### Trimmer (also referred to as "Linker")
With Xamarin apps, we could set `AndroidLinkMode`/`MtouchLink` to `Full` to remove as much unused code as possible while being able to keep specific items using an _linker-exclusion_ file.

**How can we have that package size optimization on .Net 6 mobile apps?**

### Profile Guided AOT (Android)
With Xamarin apps, we could generate an AOT profile to have the best tradeoff between package size and speed. The profiling phase would allow choosing which parts of the application to optimize to avoid having a package that's too big.

**How can we have that package size & speed optimization on .Net 6 mobile apps?**

These optimizations are crucial for enterprise-level-quality apps.

## Concern?

- [x] Usage in industry
- [x] Clarification of capabilities
- [ ] Getting started with Uno
- [x] Developing with Uno
- [ ] Contributing to the Uno project
- [x] Publishing your application
- [ ] Support
- [x] Other (please specify): Performance

## For which Platform:

- [ ] WebAssembly
- [x] Android
- [x] iOS
- [ ] macOS (AppKit)
- [ ] Mac Catalyst
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Linux Framebuffer
  - [ ] Tizen
- [ ] Windows

**Comment by jeromelaban** [comment]
The Linker (trimmer) is working as expected, and in the case mentioned in https://github.com/unoplatform/uno/discussions/11214, this is specifically related to an issue in the runtime (https://github.com/dotnet/runtime/issues/74141). 

There's no solution specified in the related changes for that issue, other than saying that generated serializers should be used. The generated serializers solution should be favored for performance reasons.

A workaround can be used though, which is to add the following in the csproj:

```xml
	<Target Name="_LinkerAdjust" AfterTargets="_PrepareLinking">
	        <!--  Workaround for https://github.com/dotnet/runtime/issues/81979 -->
		<PropertyGroup>
			<_ExtraTrimmerArgs>$(_ExtraTrimmerArgs) --keep-metadata all</_ExtraTrimmerArgs>
		</PropertyGroup>
	</Target>
```

This will inject the appropriate options in the linker so that the original .NET issue is not raised. I've opened https://github.com/dotnet/runtime/issues/81979 so that we can track the progress in the runtime.

Note that the linker configuration file should be included as mentioned in the [.NET Trimmer documentation](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0) (using `TrimmerRootDescriptor`), not using legacy Xamarin based MSBuild properties and items.

**Comment by Soap-141** [comment]
Hello @jeromelaban,

I used the workaround you mentioned and it seems to be working except that the application size is the same as the one when the linker is disabled.

Is it possible that it had nothing to remove?

![image](https://user-images.githubusercontent.com/57812285/218484967-ab20f7ae-5e76-45ce-953c-6826d98f7634.png)
![image](https://user-images.githubusercontent.com/57812285/218485579-ab9b2784-326c-4294-a2ab-714a8ee50c5a.png)
![image](https://user-images.githubusercontent.com/57812285/218485668-cb35cbb5-4cde-4a74-8774-c75a79b6ca3d.png)

**Comment by jeromelaban** [comment]
The linker is working properly:
- `<AndroidLinkMode>None</AndroidLinkMode>`
  ![image](https://user-images.githubusercontent.com/5839577/218505856-b1aa8f7f-6f95-4650-872c-0b93833fd6a9.png)
-`<AndroidLinkMode>Full</AndroidLinkMode>`
  ![image](https://user-images.githubusercontent.com/5839577/218505797-1a1a2854-541d-44a1-994c-c8b2aaa67341.png)

**Comment by jeanplevesque** [comment]
@jeromelaban, @Soap-141 tried with `PublishTrimmed` and `TrimMode`, not with `AndroidLinkMode`.
@Soap-141 Does the workaround work though? Is the app working correctly?

**Comment by jeromelaban** [comment]
It may be that `AndroidLinkMode` must not be mixed with `TrimMode`, that part is not clear.

**Comment by Soap-141** [comment]
I did not use `AndroidLinkMode` I followed this [here](https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/OneDotNet.md#linker-illink).

**Comment by Soap-141** [comment]
> @jeromelaban, @Soap-141 tried with `PublishTrimmed` and `TrimMode`, not with `AndroidLinkMode`. @Soap-141 Does the workaround work though? Is the app working correctly?

Yes it does work when using PublishTrimmed and TrimMode with the workaround but I don't think it impacted the size.

**Comment by jeanplevesque** [comment]
> Yes it does work when using PublishTrimmed and TrimMode with the workaround but I don't think it impacted the size.

You can unzip the resulting package to better see the difference, if any.

**Comment by jeromelaban** [comment]
Using the workaround above only gives a very small size difference, as it's only for method parameters.

**Comment by jeromelaban** [comment]
In recent updates, @Soap-141 mentioned that net7 is helping for this issue, and it's likely that net8 will help some more. Overall, Uno can't do much around linking itself, as those are .NET features. If we notice that specific parts of an app should be trimmed, and are not, it may be interesting to raise an issue directly in dotnet/runtime or dotnet/sdk.

What we can do in Uno, however, is enhance the size reduction with Xaml Trimming in https://github.com/unoplatform/uno/issues/9632.

**Comment by jeanplevesque** [comment]
> In recent updates, @Soap-141 mentioned that net7 is helping for this issue, and it's likely that net8 will help some more. Overall, Uno can't do much around linking itself, as those are .NET features. If we notice that specific parts of an app should be trimmed, and are not, it may be interesting to raise an issue directly in dotnet/runtime or dotnet/sdk.

But that's probably because @Soap-141 disabled AOT as a workaround for other problems. 😕

**Comment by Soap-141** [comment]
But it was still better without disabling the AOT, we just an issue in the application.

**Comment by Guidemarcus** [comment]
fyi, I know the issue is talking about net6, but while updating to net7, the same behavior is observed. 

The fix is the same, but the TrimMode property needs to be set to "partial" instead of "link"

**Comment by jeromelaban** [comment]
@Guidemarcus when saying "the fix the same", which fix are you talking about?
